ha/setup/pre_disruptive_upgrade.py

```diff
@@ -0,0 +1,115 @@
+from xml.etree.ElementTree import Element
+from ha.const import BACKUP_SOURCE_DIR, BACKUP_DEST_DIR, RA_LOG_DIR
+
+# TODO: add docstrings ones approach is settled
+
+# FIXME: Why do we need Hare directory in original script?
+HARE_DIR="/var/lib/hare"
+
+
+class PreRequisiteUpgradeError(Exception):
+    """
+    Base class for any exceptions in this module.
+    """
+
+def backup_consul(backup_dir: str) -> None:
+    Log.info("Consul backup")
+    consul_kv_dump = os.path.join(backup_dir, "consul-kv-dump.json")
+
+    if os.path.exists(consul_kv_dump):
+        timestamp = strftime("%Y%m%d%H%M%S", gmtime())
+        archive_name = f"{consul_kv_dump}.{timestamp}.tar.gz"
+        with tarfile.open(archive_name, "w:gz") as tar:
+            tar.add(consul_kv_dump)
+
+    consul_export_cmd = "consul kv export > {}".format(shlex.quote(consul_kv_dump))
+    cp = subprocess.run(consul_export_cmd, shell=True, stderr=subprocess.PIPE)
+    if cp.returncode:
+        raise PreRequisiteUpgradeError("Consul export failed with error {}".format(cp.stderr.decode()))
+
+
+def backup_configuration(src: str = BACKUP_SOURCE_DIR, dst: str = BACKUP_DEST_DIR) -> None:
+    Log.info("Taking the backup of HA configuration")
+    try:
+        if os.path.exists(dst):
+            rmtree(dst)
+        os.makedirs(dst)
+        copytree(src, dst)
+    except:
+        raise PreRequisiteUpgradeError("Failed to create backup of HA config")
+
+
+def cluster_standby_mode() -> None:
+    standby_cmd="pcs node standby --all --wait=600"
+    try:
+        SimpleCommand().run_cmd(standby_cmd)
+    except:
+        raise PreRequisiteUpgradeError("Cluster standby operation failed")
+
+
+def get_cib_xml(filename: str = None) -> Element:
+    """
+    Call `pcs cluster cib` and return XML object for further parsing
+    """
+    if filename:
+        return ElementTree.parse(filename).getroot()
+    else:
+        output, _, _ = SimpleCommand().run_cmd("pcs cluster cib")
+        return ElementTree.fromstring(output.decode())
+
+
+def delete_resources() -> None:
+    # Get the list of resources
+    try:
+        root = get_cib_xml()
+        resources = [e.attrib["id"] for e in root.findall(".//lrm_resource") 
+                     if "id" in e.attrib]
+        Log.info(f"Going to delete following resources: {resources}")
+
+        for r in resources:
+            Log.info(f"Deleting {r}")
+            SimpleCommand().run_cmd(f"pcs resource delete {r}")
+    except Exception as err:
+        raise PreRequisiteUpgradeError("Resource deletion failed", err)
+
+
+def _parse_arguments() -> Any:
+    """
+    Parse and validate input arguments passed by mini-provisioner or CLI.
+
+    Returns argparse namespace
+    """
+    parser = argparse.ArgumentParser(description="Does pre-upgrade routines")
+    parser.add_argument("--backup-consul", action="store_true", help="Do consul backup")
+    return parser.parse_args()
+
+def _main() -> None:
+    args = _parse_arguments()
```

Comment:
It will be good to log the args.

Reply:
Is this script supposed to be provided (bundled) as normal executable by cortx-ha rpm? I modified pyinstaller spec file to do that but not sure that it is actually needed.
My initial impression was that it will be used as python module and main entrypoint will be used solely for testing.

Reply:
It will be good both the options are available. There are changes being done in mini provisioner design and we may have to use it as executable as well in future.

Reply:
Ok. So, I will keep it as executable in pyinstaller spec.